### DISCUSSION TITLE
Support package file pre-processing e.g. for templated files

### DISCUSSION THREAD

**Discussion by rarkins** [renovatebot/renovate#24806] [Suggest an Idea]
**What would you like Renovate to be able to do?**

Handle/parse package files that are in template form or with something preventing them from correctly parsing (JSON, YAML, TOML, etc).

**Did you already have any implementation ideas?**

Support a `packageFilePreProcess` object, e.g. like:

```json
{
  "packageFilePreProcess": {
    "handlebars": { "name": "dummy" },
    "env": { "PACKAGE_NAME": "dummy" },
    "regex": { "<replace this>": "" }
  }
}
```

In the above example, `handlebars`, `env` and `regex` are all pre-define capabilities we support, e.g. to do a handlebars compile, or env substitution, or regex replacement.

One problem we can't easily prevent is if people use this to substitute in *dependency values* that we then detect and attempt to Renovate.

**Are there any workarounds or alternative ideas you've tried to avoid needing this feature?**

Today Renovate will complain about malformed file. For some package managers we use manual parsers instead of off-the-shelf. And sometimes people write their own regex manager.

**Is this a feature you'd be interested in implementing yourself?**

Maybe

**Comment by rarkins** [comment]
Another use case for templated helm files: https://github.com/renovatebot/config-help/issues/996

**Comment by viceice** [comment]
maybe [go-template-parser](https://github.com/bouk/go-template-parser#readme) can be a starter for go templates in yaml files

* https://golang.org/pkg/text/template/ - go template parser / processing
* https://github.com/gopherjs/gopherjs - go to js compiler

**Comment by viceice** [comment]
https://github.com/golang/go/wiki/WebAssembly

**Comment by rarkins** [comment]
We may just be better using a CLI binary

**Comment by hasanwhitesource** [comment]
@rarkins @viceice We want something new that parses these kind of files as in issue #13604, or some kind of pre-processing on the file that will work with the current functionality ?

**Comment by rarkins** [comment]
We want to add a pre processing feature so that we can keep using regular parsing in managers

**Comment by hasanwhitesource** [comment]
@rarkins @viceice What do you think if we create a new optional function in the ManagerApi interface that takes the `content` of the packageFile and does the preprocessing on it whenever some preprocessing is needed for a packageFile, and then returns the content after processing.

**Comment by rarkins** [comment]
Anything we implement should be done generically and not tied to any particular manager. The pre-processing as described above is generic: handlebars compilation, env substitution, and regex replaces.

One challenge is that `extractPackageFile` [takes content as an argument](https://github.com/renovatebot/renovate/blob/f72517dc2fc3a0ce1254749d4164d22783329a9b/lib/manager/index.ts#L61) while `extractAllPackageFiles` [takes the list of filenames](https://github.com/renovatebot/renovate/blob/f72517dc2fc3a0ce1254749d4164d22783329a9b/lib/manager/index.ts#L42) and reads content itself.

If we keep that approach, it would mean that the manager index could perform the pre-processing automatically whenever `extractPackageFile` is called, while each manager which supports `extractAllPackageFiles` would need to call the pre-processing function itself after reading content.
